src/commands/run.rs

```diff
@@ -220,79 +236,40 @@ impl RunCommand {
         result
     }
 
-    fn instantiate_module(
-        store: &Store,
-        module_registry: &ModuleRegistry,
-        path: &Path,
-    ) -> Result<Instance> {
-        // Read the wasm module binary either as `*.wat` or a raw binary
-        let data = wat::parse_file(path)?;
-
-        let module = Module::new(store, &data)?;
-
-        // Resolve import using module_registry.
-        let imports = module
-            .imports()
-            .map(|i| {
-                let export = match i.module() {
-                    "wasi_snapshot_preview1" => {
-                        module_registry.wasi_snapshot_preview1.get_export(i.name())
-                    }
-                    "wasi_unstable" => module_registry.wasi_unstable.get_export(i.name()),
-                    other => bail!("import module `{}` was not found", other),
-                };
-                match export {
-                    Some(export) => Ok(export.clone().into()),
-                    None => bail!(
-                        "import `{}` was not found in module `{}`",
-                        i.name(),
-                        i.module()
-                    ),
-                }
-            })
-            .collect::<Result<Vec<_>, _>>()?;
-
-        let instance = Instance::new(&module, &imports)
-            .context(format!("failed to instantiate {:?}", path))?;
-
-        Ok(instance)
-    }
-
-    fn handle_module(&self, store: &Store, module_registry: &ModuleRegistry) -> Result<()> {
+    fn load_main_module(&self, linker: &mut Linker) -> Result<()> {
         if let Some(timeout) = self.wasm_timeout {
```

Comment:
This is a "preexisting" bug but we should probably at some point spawn this thread before we start instantiating anything to account for `start` functions and `_initialize` and such.

(although ideally still discounting compile times of `Module`)

Reply:
Anyway, not something to handle in this PR, just a thought.

Reply:
Good point. Perhaps, if caching is enabled and a timeout is requested, we can prime the cache with module compilations, and then start the timer before we start the `Linker`. If caching is disabled, then we just count compilation time toward the timeout, perhaps.